### PR TITLE
Support AMP training through Apex

### PR DESCRIPTION
**Pull request by mats-claassen**
This PR adds support for training with Apex. 
It also makes a small modification to unfreeze the BiFPN part when specifying `--head-only`

Distributed Parallel training with Apex is not implemented

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -58,6 +64,9 @@ def get_args():
     parser.add_argument('--saved_path', type=str, default='logs/')
     parser.add_argument('--debug', type=bool, default=False, help='whether visualize the predicted boxes of trainging, '
                                                                   'the output images will be in test/')
+    parser.add_argument('--apex', help='Train using AMP with Apex', action='store_true')
+    parser.add_argument('--opt_level', type=str, default='O1', help='Opt level for Apex AMP')
+    parser.add_argument('--clip_grad', type=float, default=10.0, help='Clip gradient norm used by clip_grad_norm_')
```
Comment (mats-claassen) [inline train.py:69]: These three parameters are used when training with apex. Not specifying any of them will allow training as usual

File: train.py [hunk 2]
```diff
@@ -123,6 +132,18 @@ def train(opt):
     model = EfficientDetBackbone(num_classes=len(params.obj_list), compound_coef=opt.compound_coef,
                                  ratios=eval(params.anchors_ratios), scales=eval(params.anchors_scales))
 
+    if opt.optim == 'adamw':
+        optimizer = torch.optim.AdamW(model.parameters(), opt.lr)
+    else:
+        optimizer = torch.optim.SGD(model.parameters(), opt.lr, momentum=0.9, nesterov=True)
+
+    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, patience=3, verbose=True)
```
Comment (mats-claassen) [inline train.py:140]: These lines above were moved here from below. We need the optimizer to be defined before initializing `amp`

File: train.py [hunk 3]
```diff
@@ -135,7 +156,13 @@ def train(opt):
             last_step = 0
 
         try:
-            ret = model.load_state_dict(torch.load(weights_path), strict=False)
+            checkpoint = torch.load(weights_path)
+            if opt.apex and 'model' in checkpoint:
```
Comment (mats-claassen) [inline train.py:160]: If we are training with apex then we check whether the checkpoint includes `model`, `optimizer` and `amp` keys, which would mean it has already been trained with Apex. If it does not include these keys then we must load it as a normal checkpoint

File: train.py [hunk 4]
```diff
@@ -151,7 +178,7 @@ def train(opt):
     if opt.head_only:
         def freeze_backbone(m):
             classname = m.__class__.__name__
-            for ntl in ['EfficientNet', 'BiFPN']:
+            for ntl in ['EfficientNet']:
```
Comment (mats-claassen) [inline train.py:181]: We also want to retrain BiFPN

File: train.py [hunk 5]
```diff
@@ -230,7 +250,13 @@ def freeze_backbone(m):
                     if loss == 0 or not torch.isfinite(loss):
                         continue
 
-                    loss.backward()
+                    if opt.apex:
+                        with amp.scale_loss(loss, optimizer) as scaled_loss:
+                            scaled_loss.backward()
+                        torch.nn.utils.clip_grad_norm_(amp.master_params(optimizer), opt.clip_grad)
```
Comment (mats-claassen) [inline train.py:256]: Gradients clipped to avoid endless overflow
